ast/compile.go

```diff
@@ -114,6 +152,222 @@ func (c *Compiler) FlattenErrors() string {
+
+				// Update "safe" map to include variables in this expression
+				// that are safe, i.e., variables in the target positions
+				// of built-ins or variables in references.
+				if !e.Negated {
+					switch ts := e.Terms.(type) {
+					case *Term:
+						if r, ok := ts.Value.(Ref); ok {
+							cv := &collectAllVarsVisitor{}
+							for _, t := range r[1:] {
+								t.Walk(cv.Walk)
+							}
+							for _, v := range cv.Vars {
+								safe[v] = struct{}{}
+							}
+						}
+					case []*Term:
+						b := BuiltinMap[ts[0].Value.(Var)]
+						for i, t := range ts[1:] {
+							if t.IsGround() {
+								continue
+							}
+							if b.UnifiesRecursively(i) {
+								cv := &collectAllVarsVisitor{}
+								// If the term is a reference, any variables
+								// in the reference (except the head) are safe.
+								// Make a copy of the term so that we can exclude
+								// the head while walking the rest of the reference.
+								tmp := *t
+								if r, ok := t.Value.(Ref); ok {
+									tmp.Value = r[1:]
+								}
+								tmp.Walk(cv.Walk)
+								for _, v := range cv.Vars {
+									safe[v] = struct{}{}
+								}
+								continue
+							}
+							if v, ok := t.Value.(Var); ok {
+								if b.Unifies(i) {
+									safe[v] = struct{}{}
+									continue
+								}
+							}
+						}
+					}
+				}
+
+				// Update the set of unsafe variables for this expression and
+				// check if the expression is safe.
+				for v := range unsafe[e] {
+					if _, ok := safe[v]; ok {
+						delete(unsafe[e], v)
+					}
+				}
+
+				if len(unsafe[e]) == 0 {
+					reordered = append(reordered, e)
+					delete(unsafe, e)
+
+					// Check if other expressions in the body are considered safe
+					// now. If they are considered safe now, they can be added
+					// to the end of the re-ordered body.
+					for _, e := range r.Body {
+						if reordered.Contains(e) {
+							continue
+						}
+						for v := range unsafe[e] {
+							if _, ok := safe[v]; ok {
+								delete(unsafe[e], v)
+							}
+						}
+						if len(unsafe[e]) == 0 {
+							reordered = append(reordered, e)
+							delete(unsafe, e)
+						}
+					}
+				}
+			}
+
+			if len(unsafe) != 0 {
+				vars := map[Var]struct{}{}
+				for _, vs := range unsafe {
+					for v := range vs {
+						vars[v] = struct{}{}
+					}
+				}
+				unique := []string{}
+				for v := range vars {
+					unique = append(unique, string(v))
+				}
+				sort.Strings(unique)
+				c.err("unsafe variables in %v: %v", r.Name, strings.Join(unique, ", "))
+			} else {
+				r.Body = reordered
+			}
+		}
+	}
+}
+
+// checkSafetyHead ensures that variables appearing in the head of a
+// rule also appear in the body.
+func (c *Compiler) checkSafetyHead() {
+	for _, m := range c.Modules {
+		for _, r := range m.Rules {
+
+			headVars := []*Term{}
+
+			// Accumulate variables appearing in the head of the rule.
+			// Stop as soon as the body is encountered.
+			r.Walk(func(v interface{}) bool {
```

Comment:
Moving this into a helper is probably a good idea. We need to "walk" the head because it could the values could potentially be composites, e.g., p[k] = {"foo": x, "bar": [y]} :- ...